nntrainer/graph/graph_node.h

```diff
@@ -74,5 +75,228 @@ class GraphNode {
   // virtual GraphNode &copy(const GraphNode &from) = 0;
 };
 
+/**
+ * @brief   Iterator for GraphNode which return LayerNode object upon realize
+ *
+ * @note    This does not include the complete list of required functions. Add
+ * them as per need.
+ */
+template <typename T>
+class GraphNodeIterator : public std::iterator<std::random_access_iterator_tag,
+                                               std::shared_ptr<GraphNode>> {
+  std::shared_ptr<GraphNode> *p; /** underlying object of GraphNode */
+
+public:
+  /**
+   * @brief   iterator_traits types definition
+   *
+   * @note    these are not requried to be explicitly defined now, but maintains
+   *          forward compatibility for c++17 and later
+   *
+   * @note    valute_type, pointer and reference are different from standard
+   * iterator
+   */
+  typedef std::shared_ptr<T> value_type;
```

Comment:
Shouldn't this be returning T instead? I think this will cause some confusion if this is value_type.

Normally, Below would be expected.
```cpp
Graph g;
auto node = g.begin<LayerNode>();
LayerNode n = *node;
```

Reply:
The default behavior is :
```c++
Graph g;
auto node = g.begin();
std::shared_ptr<GraphNode> n = *node;
```

The GraphIterator allows changing GraphNode with a derived version of it. If you prefer, I can change to :
```c++
Graph g;
auto node = g.begin<std::shared_ptr<LayerNode>>();
std::shared_ptr<LayerNode> n = *node;
```

However, with this, `g.begin()`/`g.begin<LayerNode>()` gives `GraphNode &`/`LayerNode &` which are both dangerous.

Reply:
Okay